### PR TITLE
Add basic `Animation` support

### LINKED ISSUE #154: Implement `Animation` and `Transaction` structs
I think the basic stuff should be implementable with CSS transitions, but there may be some limitations. Still worth trying to implement it in my opinion.

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Resolves #154.

Nothing works at the moment, it's just an empty API to get things going.

- [ ] Support `Transaction` in `Binding`
- [ ] Hook `Animation` into `Transaction`
- [ ] Add `Animatable` protocol
- [ ] Implement at least one animatable modifier in `TokamakDOM`

### CODE REVIEW

File: Sources/TokamakCore/Animation/Animation.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  public static func interactiveSpring(
+    response: Double = 0.15, 
+    dampingFraction: Double = 0.86, 
+    blendDuration: Double = 0.25
+  ) -> Animation {
+    .init()
+  }
+  
+  public static func interpolatingSpring(
+	mass: Double = 1.0,
```
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Animation/Animation.swift:54]: - :no_entry_sign: Lines should not have trailing whitespace. (`trailing_whitespace`)

File: Sources/TokamakCore/Animation/Animation.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    response: Double = 0.15, 
+    dampingFraction: Double = 0.86, 
+    blendDuration: Double = 0.25
+  ) -> Animation {
+    .init()
+  }
+  
+  public static func interpolatingSpring(
+	mass: Double = 1.0, 
+	stiffness: Double,
```
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Animation/Animation.swift:55]: - :no_entry_sign: Lines should not have trailing whitespace. (`trailing_whitespace`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L27** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L32** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L34** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L36** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L38** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L39** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L46** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L47** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L52** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L56** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L61** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L73** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L85** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L89** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L93** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L95** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L96** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L97** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L98** - Lines should not have trailing whitespace. (`trailing_whitespace`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Animation.swift#L103** - Files should have a single trailing newline. (`trailing_newline`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Animation/Transaction.swift#L22** - Files should have a single trailing newline. (`trailing_newline`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 0ce484c77626e88be68ef96130d0a8c3c1e1c17b
</p>

**Comment by carson-katri** [comment]
The timing curves for the different easing animations are [documented on MDN](https://developer.mozilla.org/en-US/docs/Web/API/EffectTiming/easing), which seem to be the same as the [SwiftUI timing curves](https://www.objc.io/blog/2019/09/26/swiftui-animation-timing-curves/).

Springs will require a bit more calculus though 😄

**Comment by carson-katri** [comment]
I've got a spring solver working ([some great info in this article](https://medium.com/ios-os-x-development/demystifying-uikit-spring-animations-2bb868446773)). It gets pretty close. The animation could be generated to CSS keyframes or we could try fitting a cubic bezier to the curve.

```swift
struct SpringSolver {
    let ƛ: CGFloat
    let w0: CGFloat
    let wd: CGFloat
    /// Initial velocity
    let v0: CGFloat
    /// Target value
    let s0: CGFloat = 1
    
    init(mass: CGFloat, stiffness: CGFloat, damping: CGFloat, initialVelocity: CGFloat) {
        ƛ = (damping * 0.75) / (mass * 2)
        w0 = sqrt(stiffness / 2)
        wd = sqrt(abs(pow(w0, 2) - pow(ƛ, 2)))
        v0 = initialVelocity
    }
    
    func solve(at t: CGFloat) -> CGFloat {
        let y: CGFloat
        if ƛ < w0 {
            y = pow(CGFloat(M_E), -(ƛ * t)) * ((s0 * cos(wd * t)) + ((v0 + s0) * sin(wd * t)))
//      } else if ƛ > w0 { // Skip overdamping
        } else {
            y = pow(CGFloat(M_E), -(ƛ * t) * (s0 + ((v0 + (ƛ * s0)) * t)))
        }
        return 1 - y
    }
}
```

Screen Shot 2020-08-08 at 2 36 22 PM

The first parameter in the screenshots is stiffness, second is damping. `mass` is always 1 and `initialVelocity` is always 0.

**Comment by j-f1** [comment]
Very cool! It looks like the SwiftUI versions are horizontally compressed somehow, maybe?

**Comment by carson-katri** [comment]
I hard coded the length of the animation instead of waiting for it to settle before ending, so maybe that has something to do with it? Also I multiply the damping by 0.75, which may need some tweaking to look correct.

**Comment by carson-katri** [comment]
Here's what I came up with for `spring` (not `interpolatingSpring`):
```swift
static func spring(response: Double = 0.55,
                   dampingFraction: Double = 0.825,
                   blendDuration: Double = 0) -> Curve {
    if response == 0 { // Infinitely stiff spring
        return interpolatingSpring(stiffness: 999, damping: 999) // (well, not .infinity, but a very high number)
    } else {
        return interpolatingSpring(mass: 1,
                                   stiffness: pow(2 * .pi / CGFloat(response), 2),
                                   damping: 4 * .pi * CGFloat(dampingFraction) / CGFloat(response))
    }
}
```

Screen Shot 2020-08-08 at 10 31 25 PM

This time I waited for it to settle before ending the plot (`velocity < 0.01 && abs(y - 1) < 0.01`)

It should be possible to pass the generated points (and duration) to the [WebAnimation API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API/Keyframe_Formats).

**Comment by carson-katri** [comment]
Here's the Playground I made this in: https://gist.github.com/carson-katri/4cb2b963bc23bc0d87f31712e23ceb01

I added proper resting point calculation to the solver, and cleaned up the UI a bit. Again, it doesn't match SwiftUI perfectly, but it gets pretty close, and with some tweaking could probably be nearly identical.

**Comment by MaxDesiatov** [comment]
Closed as superseded by #427

**Pull request closed without merging**